### PR TITLE
Template for generating inference service domain

### LINKED ISSUE #2046: Ability to use template for generating Inference Service URL when using raw deployment
/kind feature

**Describe the solution you'd like**

Currently, inference service URL when using raw deployment is generated using hardcoded template.
```
{{ Name }}-{{ Namespace }}.{{ Domain }}
```
In some organisation, these template might need to be customised (for example at Gojek, we are using `{{ Name }}.{{ Namespace }}.{{ Domain }}` template (using dot to separate Name and Namespace), hence it should be possible for kserve admin to specify their own template. 

In the template it should be possible to use:
- Name (Inference service name)
- Namespace
- Domain ( https://github.com/kserve/kserve/blob/7464e0c0ea60279ae0e592dba2a46800e34a8c5f/pkg/apis/serving/v1beta1/configmap.go#L121 )
- Annotations
- Labels

### PR DESCRIPTION
**Pull request by pradithya**
**What this PR does / why we need it**:
This PR introduce `domainTemplate` to `ingress` configuration. It's applicable only for RawDeployment. 
Using `domainTemplate`, kserve administrator can control the domain/url generated for each inference service by combining variable from: 
1. Name of the inference service ( `{{ .Name}}` )
2. Namespace (`{{ .Namespace }}` )
3. Annotation (`{{ .Annotations.key }}`)
4. Label (`{{ .Labels.key }}`)
5. IngressDomain (`{{ .IngressDomain }}`)

The default behaviour is maintained, i.e. it uses following template: `{{ .Name }}-{{ .Namespace }}.{{ IngressDomain }}` 

**Which issue(s) this PR fixes** :
Fixes #2046

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Tested on dev cluster with default and non default `domainTemplate`. The non-default template is `{{ .Name }}.{{ .Namespace }}.{{ IngressDomain }}` which produce following ingress (notice . is separating inference service name and namespace)
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: xgboost-sample-413
  namespace: test
spec:
  rules:
  - host: xgboost-sample-413.test.example.com
    http:
      paths:
      - backend:
          service:
            name: xgboost-sample-413-predictor-default
            port:
              number: 80
        path: /
        pathType: Prefix
  - host: xgboost-sample-413-predictor-default.test.example.com
    http:
      paths:
      - backend:
          service:
            name: xgboost-sample-413-predictor-default
            port:
              number: 80
        path: /
        pathType: Prefix
```

**Special notes for your reviewer**:

1. Can't find best location to put documentation for the configuration.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add `domainTemplate` configuration for RawDeployment to configure inference service URL generation
```

### CODE REVIEW

**Comment by pradithya** [comment]
@yuzisun @Iamlovingit  Ready to be reviewed

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/kube_ingress_reconciler.go [hunk 1]
```diff
@@ -116,19 +122,17 @@ func generateMetadata(isvc *v1beta1api.InferenceService,
...
-	}
+	topLevelFlag bool) (string, error) {
+
 	metadata := generateMetadata(isvc, constants.InferenceServiceComponent(componentType))
 	if !topLevelFlag {
-		return metadata.Name + "-" + metadata.Namespace + "." + ingressConfig.IngressDomain
+		return GenerateDomainName(metadata.Name, isvc.ObjectMeta, ingressConfig)
 	} else {
-		return isvc.Name + "-" + isvc.Namespace + "." + ingressConfig.IngressDomain
+		return GenerateDomainName(isvc.Name, isvc.ObjectMeta, ingressConfig)
```
Comment (Iamlovingit) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/kube_ingress_reconciler.go:135]: It is better to add a check here

Comment (pradithya) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/kube_ingress_reconciler.go:135]: @Iamlovingit what check do you mean?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+// GenerateDomainName generate domain name using template configured in IngressConfig
+func GenerateDomainName(name string, obj metav1.ObjectMeta, ingressConfig *v1beta1.IngressConfig) (string, error) {
+	values := DomainTemplateValues{
+		Name:          name,
+		Namespace:     obj.Namespace,
+		IngressDomain: ingressConfig.IngressDomain,
+		Annotations:   obj.Annotations,
+		Labels:        obj.Labels,
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain.go:29]: Are we using labels and annotations with the template?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	tpl, err := template.New("domain-template").Parse(ingressConfig.DomainTemplate)
+	if err != nil {
+		return "", err
+	}
+
+	buf := bytes.Buffer{}
+	if err := tpl.Execute(&buf, values); err != nil {
+		return "", fmt.Errorf("error generating domain tempalte: %w", err)
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain.go]:
```suggestion
		return "", fmt.Errorf("error rendering the domain template: %w", err)
```

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain.go (hunk 2 above)
Comment (pradithya) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain.go:29]: Yes it's possible https://github.com/kserve/kserve/blob/8532e8d609cfb04e135ee7316c9021fa684e7fe7/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/domain_test.go#L66. I am making it the same as knative templating capability

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/kube_ingress_reconciler.go (hunk 1 above)
Comment (Iamlovingit) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/kube_ingress_reconciler.go:135]: > @Iamlovingit what check do you mean?

@pradithya that's fine, I am missing the change of return value.

**Comment by pradithya** [comment]
Anymore feedback @yuzisun @Iamlovingit ?

**Comment by Iamlovingit** [comment]
> Anymore feedback @yuzisun @Iamlovingit ?
Hi, @pradithya , nice job
/lgtm

**Comment by yuzisun** [comment]
Thanks @pradithya! Awesome work!

Can you help update the doc for domain template along with your previous PR ?
https://github.com/kserve/website/pull/113

**Pull request merged**